src/CBLURLEndpointListener.cc

```diff
@@ -0,0 +1,216 @@
+        if (identity->privateKey()) {
+            tls.privateKeyRepresentation = kC4PrivateKeyFromKey;
+            tls.key = identity->privateKey()->c4KeyPair();
+        } else {
+            tls.privateKeyRepresentation = kC4PrivateKeyFromCert;
+        }
+        tls.requireClientCerts = false;
+        c4config.tlsConfig = &tls;
+    }
+
+    if (_conf.authenticator) {
+        if (_conf.authenticator->withCert) {
+            // certificate
+            // Pre-condition: !_conf.disableTLS
+            tls.requireClientCerts = true;
+            tls.tlsCallbackContext = this;
+            tls.certAuthCallback = [](C4Listener *listener, C4Slice clientCertData, void *context) -> bool {
+                auto me = reinterpret_cast<CBLURLEndpointListener*>(context);
+                return me->_conf.authenticator->certCallback(me->_conf.context, clientCertData);
+            };
+        } else {
+            // user/password
+            c4config.httpAuthCallback = [](C4Listener* listener, C4Slice authHeader, void* context) -> bool {
+                slice header{authHeader};
+                auto space = header.findByte(' ');
+                if (!space) return false;
+
+                else header = header.from(space + 1);
+                while ( !header.empty() && header[0] == ' ' )
+                    header = header.from(1);
+                alloc_slice decoded = fleece::base64::decode(header);
+                auto colon = decoded.findByte(':');
+                if (!colon) return false;
+
+                slice usr = decoded.upTo(colon);
+                slice pwd = decoded.from(colon + 1);
+
+                auto me = reinterpret_cast<CBLURLEndpointListener*>(context);
+                Assert(me->_c4listener == listener);
+                return me->_conf.authenticator->pswCallback(me->_conf.context, usr, pwd);
+            };
+            c4config.callbackContext = this;
+        }
+    }
+
+    std::unique_ptr<C4Listener> c4listener{new C4Listener(c4config)};
+    auto ret = [&](bool succ) -> bool {
+        if (succ) {
+            _c4listener = c4listener.release();
+        }
+        return succ;
+    };
+
+    CBLDatabase* cblDb = _conf.collections[0]->database();
+    bool succ = true;
+    cblDb->c4db()->useLocked([&](C4Database* db) {
+        slice dbname = db->getName();
+        if ( (succ = c4listener->shareDB(dbname, db)) ) {
+            for (unsigned i = 0; i < _conf.collectionCount; ++i) {
+                _conf.collections[i]->useLocked([&](C4Collection* coll) {
+                    succ = c4listener->shareCollection(dbname, coll);
+                });
+                if (!succ) break;
+            }
+        }
+    });
+
+    return ret(succ);
+}
+
+void CBLURLEndpointListener::stop() {
+    std::scoped_lock lock(_mutex);
+    if (_c4listener) {
+        delete _c4listener;
+        _c4listener = nullptr;
+    }
+}
+
+std::string CBLURLEndpointListener::dumpConfig() {
+    return {};
```

Comment:
Do we need this function?

Reply:
Oh, I forgot to fill lit. It's for logging. I will do it next time